languages/python/django-oso/django_oso/partial.py

```diff
@@ -48,62 +51,98 @@ def partial_to_query_filter(partial, type_name):
-    assert isinstance(expression, Expression)
-    if expression.operator in COMPARISONS:
-        return compare_expr(expression, type_name)
-    elif expression.operator == "And":
-        return and_expr(expression, type_name)
-    elif expression.operator == "Isa":
-        try:
-            assert expression.args[1].tag == type_name
-        except (AssertionError, IndexError, AttributeError, TypeError):
-            raise UnsupportedError(f"Unimplemented partial isa operation {expression}.")
-
+def translate_expr(expr: Expression, model: Model, **kwargs):
+    """Translate a Polar expression to a Django Q object."""
+    assert isinstance(expr, Expression), "expected a Polar expression"
+
+    if expr.operator in COMPARISONS:
+        return compare_expr(expr, model, **kwargs)
+    elif expr.operator == "And":
+        return and_expr(expr, model, **kwargs)
+    elif expr.operator == "Isa":
+        for attr in dot_op_path(expr.args[0]):
+            model = getattr(model, attr).field.related_model
+        constraint_type = apps.get_model(django_model_name(expr.args[1].tag))
+        if not issubclass(constraint_type, model):
+            raise UnexpectedPolarTypeError(
+                f"Type constraint violation on partial.\nConstraint: {polar_model_name(constraint_type)}; Field: {polar_model_name(model)}"
+            )
         return None
+    elif expr.operator == "In":
+        return in_expr(expr, model, **kwargs)
     else:
-        raise UnsupportedError(f"Unimplemented partial operator {expression.operator}")
+        raise UnsupportedError(f"Unimplemented partial operator {expr.operator}")
 
 
-def and_expr(expr, type_name):
+def and_expr(expr: Expression, model: Model, **kwargs):
+    assert expr.operator == "And"
     q = Q()
+    for arg in expr.args:
+        expr = translate_expr(arg, model, **kwargs)
+        if expr:
+            q = q & expr
+    return q
 
-    assert expr.operator == "And"
-    for expression in expr.args:
-        expr = translate_expr(expression, type_name)
-        if expr is None:
-            continue
 
-        q = q & expr
+def compare_expr(expr: Expression, _model: Model, path=[], **kwargs):
+    q = Q()
+    (left, right) = expr.args
+    left_path = dot_op_path(left)
+    if left_path:
+        return COMPARISONS[expr.operator](q, "__".join(path + left_path), right)
+    else:
+        if isinstance(right, Model):
+            right = right.pk
+        else:
+            raise UnsupportedError(f"Unsupported comparison: {expr}")
+        return COMPARISONS[expr.operator](q, "__".join(path + ["pk"]), right)
 
-    return q
 
+def in_expr(expr: Expression, model: Model, path=[], **kwargs):
+    assert expr.operator == "In"
+    q = Q()
+    (left, right) = expr.args
+    right_path = dot_op_path(right)
+    assert right_path, "RHS of in must be a dot lookup"
+    right_path = path + right_path
+
+    if isinstance(left, Expression):
+        if left.operator == "And":
```

Comment:
Why can't this just call `translate_expr(left, model, path=right_path, **kwargs)`?

Reply:
826028b